### PR TITLE
feat: fill in proof of Array.data_erase

### PR DESCRIPTION
**Pull request by Seppel3210**
I also added some lemmas in List.Lemmas that were necessary.
I'm not sure about the naming of some of the lemmas I added, so I would appreciate feedback!

### CODE REVIEW

**Comment by Seppel3210** [comment]
awaiting-review

**Comment by Seppel3210** [comment]
Oops 🤦  Linter errors should be fixed now

File: Std/Data/Array/Lemmas.lean [hunk 1]
```diff
@@ -721,6 +721,87 @@ theorem all_def {p : α → Bool} (as : Array α) : as.all p = as.data.all p :=
 theorem all_eq_true_iff_forall_mem {l : Array α} : l.all p ↔ ∀ x, x ∈ l → p x := by
   simp only [all_def, List.all_eq_true, mem_def]
 
+/-! ### indexOf? -/
+
+@[simp] theorem indexOf?_data [BEq α] {a : α} {l : Array α} :
+    (l.indexOf? a).map Fin.val = l.data.indexOf? a := by
```
Comment (semorrison) [inline Std/Data/Array/Lemmas.lean:727]: Let's not have this be `@[simp]`.

The name is fine, but only if you switch LHS and RHS in the statement. Considering how you use it below, I think that ordering would make more sense.

File: Std/Data/Array/Lemmas.lean [hunk 2]
```diff
@@ -721,6 +721,87 @@ theorem all_def {p : α → Bool} (as : Array α) : as.all p = as.data.all p :=
...
+  have h := aux l 0
+  simp at h
+  assumption
```
Comment (semorrison) [inline Std/Data/Array/Lemmas.lean:730]:
```suggestion
  simpa using aux l 0
```

File: Std/Data/Array/Lemmas.lean [hunk 3]
```diff
@@ -721,6 +721,87 @@ theorem all_def {p : α → Bool} (as : Array α) : as.all p = as.data.all p :=
...
+    rw [pop, ←List.dropLast_eq_eraseIdx]
+    simp at last_idx ⊢
+    exact last_idx.symm
+termination_by l.size - i
+
+@[simp] theorem erase_data [BEq α] {l : Array α} {a : α} : (l.erase a).data = l.data.erase a := by
+  let ⟨xs⟩ := l
+  match h : indexOf? ⟨xs⟩ a with
+  | none =>
+    simp [erase, h]
```
Comment (semorrison) [inline Std/Data/Array/Lemmas.lean:785]: Non-terminal simp. Please use `simp?` to find a `simp only` you can replace this with. Also elsewhere.

File: Std/Data/Array/Lemmas.lean [hunk 4]
```diff
@@ -721,6 +721,87 @@ theorem all_def {p : α → Bool} (as : Array α) : as.all p = as.data.all p :=
...
+
 /-! ### erase -/
 
-@[simp] proof_wanted erase_data [BEq α] {l : Array α} {a : α} : (l.erase a).data = l.data.erase a
+theorem eraseIdx_swap {l : Array α} (i : Nat) (lt : i + 1 < size l) :
+    (l.swap ⟨i+1, lt⟩ ⟨i, Nat.lt_of_succ_lt lt⟩).data.eraseIdx (i+1) = l.data.eraseIdx i := by
+  induction i generalizing l with
+  | zero =>
+    let ⟨xs⟩ := l
+    let x₀::x₁::xs := xs
```
Comment (semorrison) [inline Std/Data/Array/Lemmas.lean:753]: Can you do these in one go? (And below?)

File: Std/Data/Array/Lemmas.lean [hunk 5]
```diff
@@ -721,6 +721,87 @@ theorem all_def {p : α → Bool} (as : Array α) : as.all p = as.data.all p :=
...
+  | succ i ih =>
+    let ⟨xs⟩ := l
+    let x::xs := xs
+    have lt' := Nat.lt_of_succ_lt_succ lt
+    have : (swap ⟨x::xs⟩ ⟨i.succ + 1, lt⟩ ⟨i.succ, Nat.lt_of_succ_lt lt⟩).data
+        = x::(swap ⟨xs⟩ ⟨i + 1, lt'⟩ ⟨i, Nat.lt_of_succ_lt lt'⟩).data := by
+      simp [swap_def, List.set_succ, getElem_eq_data_get]
+    simp [this, ih]
+
+theorem eraseIdxAux_data {l : Array α} {i : Nat} (le : i+1 ≤ l.size) :
```
Comment (semorrison) [inline Std/Data/Array/Lemmas.lean:764]:
```suggestion
theorem data_eraseIdxAux {l : Array α} {i : Nat} (le : i+1 ≤ l.size) :
```

**Comment by semorrison** [comment]
Otherwise looks good.

File: Std/Data/Array/Lemmas.lean [hunk 6]
```diff
@@ -721,6 +721,87 @@ theorem all_def {p : α → Bool} (as : Array α) : as.all p = as.data.all p :=
...
+    exact erase_none 0 h
+  | some i =>
+    simp [erase, h]
+    rw [feraseIdx, eraseIdxAux_data i.is_lt, ←List.eraseIdx_indexOf_eq_erase]
+    congr
+    rw [List.indexOf_eq_indexOf?, ←indexOf?_data]
+    simp [h]
+where
+  erase_none {xs : List α} (i : Nat) (h : List.indexOf? a xs = none) :
+      xs.drop i = (xs.drop i).erase a := by
```
Comment (timotree3) [inline Std/Data/Array/Lemmas.lean:797]: Can `erase_none` be factored into these three more generally useful lemmas?
```lean
theorem erase_eq_of_nmem : x ∉ xs -> xs.erase x = xs := sorry
theorem nmem_drop_of_nmem : x ∉ xs -> x ∉ xs.drop i := sorry
theorem indexOf?_eq_none : List.indexOf? x xs = none ↔ x ∉ xs := sorry
```

**Comment by timotree3** [comment]
I thought it was a bit ugly that this PR had to unfold morally private definitions from core like `eraseIdxAux` so I made a PR which would allow us to simply unfold `feraseIdx` instead: leanprover/lean4#3676. If that PR gets merged it will break the proofs here.

**Comment by Seppel3210** [comment]
as I said on the other PR @timotree3: In the case that your PR gets merged (and std4's lean-toolchain gets bumped) I have a commit lying around that fixes the proof https://github.com/Seppel3210/std4/commit/109c3ce97e182813a166c9b55418d9a599a01bb6

**Comment by Seppel3210** [comment]
I think it makes sense to wait for the lean4 version to get bumped since the lean4 PR was merged. Or should I PR it to the bump branch instead? (@semorrison you probably know this)

**Comment by semorrison** [comment]
Yes, if you'd like to PR to `bump/v4.8.0` that's fine. It's also fine to wait: hopefully the release will be out next Tuesday.

**Comment by semorrison** [comment]
Sorry, the release was much delayed, but now should be out in the next few days again!

**Comment by Seppel3210** [comment]
Not really sure who to ping for this issue :sweat_smile: @semorrison

**Comment by semorrison** [comment]
Let's summon @nomeata.

**Comment by nomeata** [comment]
I think the problem isn’t with `feraseIdx.induct` per se, but rather with `rw [eraseIdx]`, because
```diff
diff --git a/Std/Data/Array/Lemmas.lean b/Std/Data/Array/Lemmas.lean
index 31d9a385..a56c8c5e 100644
--- a/Std/Data/Array/Lemmas.lean
+++ b/Std/Data/Array/Lemmas.lean
@@ -843,16 +843,24 @@ theorem eraseIdx_swap_data {l : Array α} (i : Nat) (lt : i + 1 < size l) :
       simp [swap_def, List.set_succ, getElem_eq_data_get]
     simp [this, ih]

+theorem feraseIdx_eq (a : Array α) (i : Fin a.size) :
+  a.feraseIdx i =
+    if h : ↑i + 1 < a.size then
+      let a' := a.swap ⟨↑i + 1, h⟩ i;
+      let i' : Fin a'.size := ⟨i.val + 1, by simp [a', h]⟩
+      a'.feraseIdx i'
+    else a.pop := by rw [feraseIdx]
+
 theorem feraseIdx_data {l : Array α} {i : Fin l.size} :
     (l.feraseIdx i).data = l.data.eraseIdx i := by
   induction l, i using feraseIdx.induct with
   | @case1 a i lt a' i' _ ih =>
-    rw [feraseIdx]
+    rw [feraseIdx_eq]
     simp [lt, ih, a', eraseIdx_swap_data i lt]
   | case2 a i lt =>
     have : i + 1 ≥ a.size := Nat.ge_of_not_lt lt
     have last : i + 1 = a.size := Nat.le_antisymm i.is_lt this
-    simp [feraseIdx, lt, List.dropLast_eq_eraseIdx last]
+    simp [feraseIdx_eq, lt, List.dropLast_eq_eraseIdx last]

 @[simp] theorem erase_data [BEq α] {l : Array α} {a : α} : (l.erase a).data = l.data.erase a := by
   let ⟨xs⟩ := l
```
moves the error there.

The problem is that `eraseIdx.eq_1` has these unnecessary `have` on the right-hand side.

The proper fix that I’d advocate for is to move everything that’s only needed for the termination argument into `decerasing_by`, this way you get the proper equational lemma.

If you don't want to wait with this PR until that lands, just mark the proof as `nolint` until this is fixed (`import Std.Tactic.Lint.Misc`, `@[nolint unusedHavesSuffices] -- explanation`).

**Comment by Seppel3210** [comment]
should be good now @semorrison

**Comment by Seppel3210** [comment]
awaiting-review

**Comment by Seppel3210** [comment]
awaiting-review

File: Batteries/Data/List/Lemmas.lean [hunk 7]
```diff
@@ -736,6 +752,17 @@ theorem findIdx?_of_eq_none {xs : List α} {p : α → Bool} (w : xs.findIdx? p
     simp only [replicate, findIdx?_cons, Nat.zero_add, findIdx?_succ, Nat.zero_lt_succ, true_and]
     split <;> simp_all
 
+theorem findIdx_eq_findIdx? (p : α → Bool) (l : List α) :
+    l.findIdx p = (match l.findIdx? p with | some i => i | none => l.length) := by
```
Comment (b-mehta) [inline Batteries/Data/List/Lemmas.lean:756]: Is this easier to state and prove with Option.elim?

File: Batteries/Data/Array/Lemmas.lean [hunk 8]
```diff
@@ -112,9 +112,77 @@ theorem mem_join : ∀ {L : Array (Array α)}, a ∈ L.join ↔ ∃ l, l ∈ L 
   · rintro ⟨s, h₁, h₂⟩
     refine ⟨s.data, ⟨⟨s, h₁, rfl⟩, h₂⟩⟩
 
+/-! ### indexOf? -/
+
```
Comment (Shreyas4991) [inline Batteries/Data/Array/Lemmas.lean:116]: I don't think you're supposed to leave a line between the docstring and the definition

File: Batteries/Data/Array/Lemmas.lean [hunk 9]
```diff
@@ -112,9 +112,77 @@ theorem mem_join : ∀ {L : Array (Array α)}, a ∈ L.join ↔ ∃ l, l ∈ L 
...
+      if h' : l[i] == a then
+        simp [h, h']
+      else
+        simp [h, h', ←aux l (i+1), Function.comp_def, ←Nat.add_assoc, Nat.add_right_comm]
+    else
+      have h' : l.size ≤ i := Nat.le_of_not_lt h
+      simp [h, List.drop_length_le h', List.indexOf?]
+  termination_by l.size - i
+
 /-! ### erase -/
```
Comment (Shreyas4991) [inline Batteries/Data/Array/Lemmas.lean:136]: Same as above. No newlines between docstring and comments.

File: Batteries/Data/Array/Lemmas.lean [hunk 10]
```diff
@@ -112,9 +112,77 @@ theorem mem_join : ∀ {L : Array (Array α)}, a ∈ L.join ↔ ∃ l, l ∈ L 
...
+  case succ i ih _ =>
+    have lt' := Nat.lt_of_succ_lt_succ lt
+    have : (swap ⟨x₀::x₁::xs⟩ ⟨i.succ + 1, lt⟩ ⟨i.succ, Nat.lt_of_succ_lt lt⟩).data
+        = x₀::(swap ⟨x₁::xs⟩ ⟨i + 1, lt'⟩ ⟨i, Nat.lt_of_succ_lt lt'⟩).data := by
+      simp [swap_def, List.set_succ, getElem_eq_data_get]
+    simp [this, ih]
+
+theorem feraseIdx_data {l : Array α} {i : Fin l.size} :
+    (l.feraseIdx i).data = l.data.eraseIdx i := by
+  induction l, i using feraseIdx.induct with
```
Comment (b-mehta) [inline Batteries/Data/Array/Lemmas.lean:151]: Nice to see `.induct` getting used! Does it help with the other proofs you have here?

File: Batteries/Data/List/Lemmas.lean [hunk 11]
```diff
@@ -241,6 +241,22 @@ theorem tail_eq_tail? (l) : @tail α l = (tail? l).getD [] := by simp [tail_eq_t
 @[simp] theorem next?_nil : @next? α [] = none := rfl
 @[simp] theorem next?_cons (a l) : @next? α (a :: l) = some (a, l) := rfl
 
+/-! ### dropLast -/
+
+theorem dropLast_eq_eraseIdx {xs : List α} {i : Nat} (last_idx : i + 1 = xs.length) :
+    xs.dropLast = List.eraseIdx xs i := by
+  induction i generalizing xs with
+  | zero =>
+    let [x] := xs
```
Comment (b-mehta) [inline Batteries/Data/List/Lemmas.lean:250]: This line looks unnecessary?

File: Batteries/Data/Array/Lemmas.lean [hunk 12]
```diff
@@ -112,9 +112,77 @@ theorem mem_join : ∀ {L : Array (Array α)}, a ∈ L.join ↔ ∃ l, l ∈ L 
...
+    rw [feraseIdx]
+    simp [lt, ih, a', eraseIdx_swap_data i lt]
+  | case2 a i lt =>
+    have : i + 1 ≥ a.size := Nat.ge_of_not_lt lt
+    have last : i + 1 = a.size := Nat.le_antisymm i.is_lt this
+    simp [feraseIdx, lt, List.dropLast_eq_eraseIdx last]
+
+@[simp] theorem erase_data [BEq α] {l : Array α} {a : α} : (l.erase a).data = l.data.erase a := by
+  let ⟨xs⟩ := l
+  match h : indexOf? ⟨xs⟩ a with
```
Comment (b-mehta) [inline Batteries/Data/Array/Lemmas.lean:162]:
```suggestion
  match h : indexOf? l a with
```

File: Batteries/Data/List/Lemmas.lean [hunk 13]
```diff
@@ -241,6 +241,22 @@ theorem tail_eq_tail? (l) : @tail α l = (tail? l).getD [] := by simp [tail_eq_t
 @[simp] theorem next?_nil : @next? α [] = none := rfl
 @[simp] theorem next?_cons (a l) : @next? α (a :: l) = some (a, l) := rfl
 
+/-! ### dropLast -/
+
```
Comment (Shreyas4991) [inline Batteries/Data/List/Lemmas.lean:245]: same as above. Please remove that newline.

File: Batteries/Data/List/Lemmas.lean [hunk 14]
```diff
@@ -1481,6 +1508,14 @@ theorem indexOf_cons [BEq α] :
   dsimp [indexOf]
   simp [findIdx_cons]
 
+@[simp] theorem eraseIdx_indexOf_eq_erase [BEq α] (a : α) (l : List α) :
```
Comment (Shreyas4991) [inline Batteries/Data/List/Lemmas.lean:1511]: The typical syntax is 
```
@[simp]
theorem <name...>
```
There is usually a newline after the attribute.

File: Batteries/Data/Array/Lemmas.lean [hunk 15]
```diff
@@ -112,9 +112,77 @@ theorem mem_join : ∀ {L : Array (Array α)}, a ∈ L.join ↔ ∃ l, l ∈ L 
...
+  induction l, i using feraseIdx.induct with
+  | @case1 a i lt a' i' ih =>
+    rw [feraseIdx]
+    simp [lt, ih, a', eraseIdx_swap_data i lt]
+  | case2 a i lt =>
+    have : i + 1 ≥ a.size := Nat.ge_of_not_lt lt
+    have last : i + 1 = a.size := Nat.le_antisymm i.is_lt this
+    simp [feraseIdx, lt, List.dropLast_eq_eraseIdx last]
+
+@[simp] theorem erase_data [BEq α] {l : Array α} {a : α} : (l.erase a).data = l.data.erase a := by
```
Comment (Shreyas4991) [inline Batteries/Data/Array/Lemmas.lean:160]: a new line after the `@[simp]` attribute is the usual style.

File: Batteries/Data/List/Lemmas.lean [hunk 16]
```diff
@@ -1495,6 +1530,15 @@ theorem indexOf_mem_indexesOf [BEq α] [LawfulBEq α] {xs : List α} (m : x ∈
         specialize ih m
         simpa
 
+@[simp] theorem indexOf?_nil [BEq α] : ([] : List α).indexOf? x = none := rfl
```
Comment (Shreyas4991) [inline Batteries/Data/List/Lemmas.lean:1533]: A new line after the `simp` attribute is the usual style. Same as above.

**Review by Shreyas4991 — COMMENTED** [review]
Thanks for the PR. I have highlighted mostly formatting issues in this pass. Nothing else seems out of order.

**Review by Shreyas4991 — APPROVED** [review]
The attribute changes are not necessary as @b-mehta just informed me. The docstring spaces are trivial to fix. Nothing else appears wrong. There are no new `defs` and hence no performance issues with this PR whatsoever.

File: Batteries/Data/Array/Lemmas.lean (hunk 12 above)
Comment (Shreyas4991) [reply Batteries/Data/Array/Lemmas.lean:162]: I am curious how this came about.

File: Batteries/Data/Array/Lemmas.lean (hunk 8 above)
Comment (b-mehta) [reply Batteries/Data/Array/Lemmas.lean:116]: Note this isn't a docstring, and it's also following the pattern of the lemmas elsewhere in this file.

File: Batteries/Data/Array/Lemmas.lean [hunk 17]
```diff
@@ -112,9 +112,77 @@ theorem mem_join : ∀ {L : Array (Array α)}, a ∈ L.join ↔ ∃ l, l ∈ L 
...
+    apply erase_none 0
+    simpa [←indexOf?_data] using congrArg (Option.map Fin.val) h
+  | some i =>
+    simp only [erase, h]
+    rw [feraseIdx_data, ←List.eraseIdx_indexOf_eq_erase]
+    congr
+    rw [List.indexOf_eq_indexOf?, indexOf?_data]
+    simp [h]
+where
+  erase_none {xs : List α} (i : Nat) (h : List.indexOf? a xs = none) :
```
Comment (b-mehta) [inline Batteries/Data/Array/Lemmas.lean:174]: It looks to me like this should be its own lemma, it seems independently useful. I also agree with the comments about rephrasing it in terms of memberships, as made above.

File: Batteries/Data/List/Lemmas.lean [hunk 18]
```diff
@@ -613,6 +613,10 @@ theorem erase_comm [LawfulBEq α] (a b : α) (l : List α) :
       rw [erase_append_right _ h₁, erase_append_right _ h₁, erase_append_right _ ha',
           erase_cons_tail _ ab, erase_cons_head]
 
+theorem erase_of_forall_bne (a : α) (xs : List α) (h : ∀ (x : α), x ∈ xs → ¬x == a) :
```
Comment (b-mehta) [inline Batteries/Data/List/Lemmas.lean:616]: If you don't want to also add the lemma that [LawfulBEq α] and `a ∉ xs` implies `xs.erase a = xs`, maybe add a TODO for that?

File: Batteries/Data/List/Lemmas.lean (hunk 7 above)
Comment (fgdorais) [reply Batteries/Data/List/Lemmas.lean:769]: Or just `Option.getD`:
```suggestion
    l.findIdx p = (l.findIdx? p).getD l.length := by
```

**Comment by Seppel3210** [comment]
not sure why this is waiting on me again… I intended to push the TODO in another PR since this one is already pretty big, but I can also do it in this one I guess.

**Comment by Seppel3210** [comment]
awaiting-review

**Comment by Seppel3210** [comment]
awaiting-review

File: Batteries/Data/Array/Lemmas.lean [hunk 19]
```diff
@@ -111,9 +111,64 @@ theorem mem_join : ∀ {L : Array (Array α)}, a ∈ L.join ↔ ∃ l, l ∈ L 
...
+        simp [h, h', ←aux l (i+1), Function.comp_def, ←Nat.add_assoc, Nat.add_right_comm]
+    else
+      have h' : l.size ≤ i := Nat.le_of_not_lt h
+      simp [h, List.drop_length_le h', List.indexOf?]
+  termination_by l.size - i
+
 /-! ### erase -/
 
-@[simp] proof_wanted erase_data [BEq α] {l : Array α} {a : α} : (l.erase a).data = l.data.erase a
+theorem eraseIdx_swap_data {l : Array α} (i : Nat) (lt : i + 1 < size l) :
```
Comment (fgdorais) [inline Batteries/Data/Array/Lemmas.lean:136]:
```suggestion
theorem eraseIdx_data_swap {l : Array α} (i : Nat) (lt : i + 1 < size l) :
```

File: Batteries/Data/Array/Lemmas.lean [hunk 20]
```diff
@@ -111,9 +111,64 @@ theorem mem_join : ∀ {L : Array (Array α)}, a ∈ L.join ↔ ∃ l, l ∈ L 
...
+  induction i generalizing xs <;> let x₀::x₁::xs := xs
+  case zero => simp [swap, get]
+  case succ i ih _ =>
+    have lt' := Nat.lt_of_succ_lt_succ lt
+    have : (swap ⟨x₀::x₁::xs⟩ ⟨i.succ + 1, lt⟩ ⟨i.succ, Nat.lt_of_succ_lt lt⟩).data
+        = x₀::(swap ⟨x₁::xs⟩ ⟨i + 1, lt'⟩ ⟨i, Nat.lt_of_succ_lt lt'⟩).data := by
+      simp [swap_def, getElem_eq_data_getElem]
+    simp [this, ih]
+
+theorem feraseIdx_data {l : Array α} {i : Fin l.size} :
```
Comment (fgdorais) [inline Batteries/Data/Array/Lemmas.lean:148]:
```suggestion
theorem data_feraseIdx {l : Array α} (i : Fin l.size) :
```

File: Batteries/Data/Array/Lemmas.lean [hunk 21]
```diff
@@ -111,9 +111,64 @@ theorem mem_join : ∀ {L : Array (Array α)}, a ∈ L.join ↔ ∃ l, l ∈ L 
...
+  induction l, i using feraseIdx.induct with
+  | @case1 a i lt a' i' ih =>
+    rw [feraseIdx]
+    simp [lt, ih, a', eraseIdx_swap_data i lt]
+  | case2 a i lt =>
+    have : i + 1 ≥ a.size := Nat.ge_of_not_lt lt
+    have last : i + 1 = a.size := Nat.le_antisymm i.is_lt this
+    simp [feraseIdx, lt, List.dropLast_eq_eraseIdx last]
+
+@[simp] theorem erase_data [BEq α] {l : Array α} {a : α} : (l.erase a).data = l.data.erase a := by
```
Comment (fgdorais) [inline Batteries/Data/Array/Lemmas.lean:159]:
```suggestion
@[simp] theorem data_erase [BEq α] (l : Array α) (a : α) : (l.erase a).data = l.data.erase a := by
```

File: Batteries/Data/List/Lemmas.lean [hunk 22]
```diff
@@ -205,6 +221,32 @@ theorem get?_set_of_lt' (a : α) {m n} (l : List α) (h : m < length l) :
...
 
+theorem erase_of_forall_bne [BEq α] (a : α) (xs : List α) (h : ∀ (x : α), x ∈ xs → ¬x == a) :
+    xs.erase a = xs := by
+  rw [erase_eq_eraseP', eraseP_of_forall_not h]
+
+-- TODO a version of the above theorem with LawfulBEq and ∉
+
+/-! ### findIdx? -/
+
+theorem findIdx?_eq_none (xs : List α) (p : α → Bool) :
```
Comment (fgdorais) [inline Batteries/Data/List/Lemmas.lean:232]:
```suggestion
theorem findIdx?_eq_none_iff {xs : List α} {p : α → Bool} :
```

File: Batteries/Data/List/Lemmas.lean [hunk 23]
```diff
@@ -562,6 +612,19 @@ theorem indexOf_mem_indexesOf [BEq α] [LawfulBEq α] {xs : List α} (m : x ∈
         specialize ih m
         simpa
 
+@[simp] theorem indexOf?_nil [BEq α] : ([] : List α).indexOf? x = none := rfl
+theorem indexOf?_cons [BEq α] :
+    (x :: xs : List α).indexOf? y = if x == y then some 0 else (xs.indexOf? y).map Nat.succ := by
+  simp [indexOf?]
+
+theorem indexOf?_eq_none_iff [BEq α] (a : α) (l : List α) :
```
Comment (fgdorais) [inline Batteries/Data/List/Lemmas.lean:620]:
```suggestion
theorem indexOf?_eq_none_iff [BEq α] {a : α} {l : List α} :
```

**Review by fgdorais — CHANGES_REQUESTED** [review]

**Comment by Seppel3210** [comment]
thanks for taking a look!

**Comment by Seppel3210** [comment]
awaiting-review

File: Batteries/Data/Array/Lemmas.lean [hunk 24]
```diff
@@ -115,9 +115,64 @@ theorem mem_join : ∀ {L : Array (Array α)}, a ∈ L.join ↔ ∃ l, l ∈ L 
   · rintro ⟨s, h₁, h₂⟩
     refine ⟨s.data, ⟨⟨s, h₁, rfl⟩, h₂⟩⟩
 
+/-! ### indexOf? -/
+
+theorem indexOf?_data [BEq α] {a : α} {l : Array α} :
```
Comment (fgdorais) [inline Batteries/Data/Array/Lemmas.lean:120]: Is there a reason this is not simp?

File: Batteries/Data/Array/Lemmas.lean [hunk 25]
```diff
@@ -115,9 +115,64 @@ theorem mem_join : ∀ {L : Array (Array α)}, a ∈ L.join ↔ ∃ l, l ∈ L 
...
+  induction i generalizing xs <;> let x₀::x₁::xs := xs
+  case zero => simp [swap, get]
+  case succ i ih _ =>
+    have lt' := Nat.lt_of_succ_lt_succ lt
+    have : (swap ⟨x₀::x₁::xs⟩ ⟨i.succ + 1, lt⟩ ⟨i.succ, Nat.lt_of_succ_lt lt⟩).data
+        = x₀::(swap ⟨x₁::xs⟩ ⟨i + 1, lt'⟩ ⟨i, Nat.lt_of_succ_lt lt'⟩).data := by
+      simp [swap_def, getElem_eq_data_getElem]
+    simp [this, ih]
+
+theorem data_feraseIdx {l : Array α} (i : Fin l.size) :
```
Comment (fgdorais) [inline Batteries/Data/Array/Lemmas.lean:152]: Is there a reason this is not simp?

**Review by fgdorais — COMMENTED** [review]
Sorry for the delay following up! I just have a couple of leftover questions.

File: Batteries/Data/Array/Lemmas.lean (hunk 24 above)
Comment (fgdorais) [reply Batteries/Data/Array/Lemmas.lean:120]: Makes sense. Thanks for looking into it.

**Review by fgdorais — APPROVED** [review]

**Pull request merged**